Add pre-training transform to convert BatchNorm to BatchNormInternal

pranav-prakash opened the pull request:
**Description**: As opset 14 lacks saved_mean/saved_inv_std, we have to convert nodes to a custom op if we want to ensure good performance on backward pass. This is the second in a series of PRs to implement BN training on CPU (first was #6946).

As part of this we also remove the existing `AdjustBatchNormOutputs` pass – this only works for opset 9 BN and is subsumed by the more general `BatchNormReplacement` we define in this PR.

WIP – I think the only thing missing is a unit test. I'd also appreciate any feedback on whether it's necessary to copy/paste the entire schema-def  for the `BatchNormInternal` op or whether something like protobuf's `toBuilder` exists for these schema defs.

SherlockNoMad commented:
Hi @pranav-prakash, thanks for contributing this ! 

Let us know when this is ready for review. I will help you merging this change.

pranav-prakash commented:
@SherlockNoMad
Thanks, I just added the test so it should be ready for review. One question I had was that the test required me to explicitly import the `com.microsoft` domain in order for it to recognize the `BatchNormInternal` op – I'm not sure if this is expected behavior. When loading a model from disk is `com.microsoft` automatically added as a valid import?

SherlockNoMad commented on orttraining/orttraining/test/optimizer/graph_transform_test.cc:
```diff
@@ -33,6 +34,49 @@ namespace test {
+  
+  ASSERT_TRUE(graph.NumberOfNodes() == 1);
+  // Make sure that BN was updated to add outputs
+  ASSERT_TRUE(graph.Nodes().begin()->MutableOutputDefs().size() == 5);
```
maybe one more assert on the optype to be BacthNormInternal?

pranav-prakash replied:
Done

SherlockNoMad commented on orttraining/orttraining/core/framework/gradient_graph_builder.cc:
```diff
@@ -32,7 +32,6 @@ GradientGraphBuilder::GradientGraphBuilder(Graph* graph,
   rule_based_graph_transformer->Register(make_unique<InsertMaxPoolOutput>());
-  rule_based_graph_transformer->Register(make_unique<AdjustBatchNormOutputs>());
 
   graph_transformation_mgr_.Register(std::move(rule_based_graph_transformer),
```
InsertMaxPoolOutput can also be removed here. 
actually, graph_transformation_mgr_ can be removed from this class.

pranav-prakash replied:
Done.

SherlockNoMad commented:
I think we also need to update gradient_builder_registry.cc, to register the gradient builder for BatchNormInternal. 
We should be able to reuse "GetBatchNormalizationGradient" though.

SherlockNoMad commented:
> @SherlockNoMad
> Thanks, I just added the test so it should be ready for review. One question I had was that the test required me to explicitly import the `com.microsoft` domain in order for it to recognize the `BatchNormInternal` op – I'm not sure if this is expected behavior. When loading a model from disk is `com.microsoft` automatically added as a valid import?

I can imagine msDomain is needed to make the UT work. In the training process, msDomain kernel registry will be automatically included at program start time, coz most of the gradient ops are under MSDomain.

SherlockNoMad commented on orttraining/orttraining/core/optimizer/batchnorm_replacement.cc:
```diff
@@ -0,0 +1,55 @@
+  auto scale_input_def_type_proto = scale_input_def->TypeAsProto();
+
+  // Guard against a BatchNorm that already has optional outputs present for some reason
+  if (bn_outputs.size() == 1) {
```
what about the case bn_outputs.size() == 3?

pranav-prakash replied:
Fixed, thanks for catching this!

SherlockNoMad commented on orttraining/orttraining/core/optimizer/batchnorm_replacement.cc:
```diff
@@ -0,0 +1,55 @@
+  // check Batch Normalization node has 5 output node args for training mode
+  ORT_ENFORCE(bn_node.OutputDefs().size() == 5);
+
+  Node& batchnorm_internal_node = graph.AddNode(graph.GenerateNodeName("BatchNormInternal"),
```
would be good to preserve the original node name as a substring of the new name
so that we can track the original of node

pranav-prakash replied:
Fixed, I'm doing `bn_node.Name() + "_BatchNormInternal"` which should hopefully suffice

SherlockNoMad commented on orttraining/orttraining/test/optimizer/graph_transform_test.cc:
```diff
@@ -33,6 +34,49 @@ namespace test {
+  auto& input_mean = graph.GetOrCreateNodeArg("input_mean", &scale_tensor_type);
+  auto& input_var = graph.GetOrCreateNodeArg("input_var", &scale_tensor_type);
+
+  auto& output_Y = graph.GetOrCreateNodeArg("Y", &input_tensor_type);
```
maybe another test case with 3 outputs?

pranav-prakash replied:
Done. Added a testcase for opset14 with 3 outputs, as well as opset 9 with 5 outputs

SherlockNoMad commented:
Also need to add BatchNormInternal into the STOP_GRADIENT_EDGES in gradient_graph_builder.h

SherlockNoMad commented:
Thanks again for the PR. The bulk part is looking good! 
We should be good to merge once the comments are addressed

pranav-prakash commented:
Thanks for the quick review! I addressed the comments you left.

>register gradient builder for BatchNormInternal.
Done. I'll add a CPU implementation of the gradient in a subsequent PR.

I'll also register an implementation for BatchNormInternal (re-using existing BatchNorm kernel) in a subsequent PR. I believe there's also a pre-existing CUDA kernel for this, but it seems to be broken (tests were disabled) so someone else who's both more familiar and capable of testing it will have to modify those parts.

SherlockNoMad commented:
/azp run Linux CPU CI Pipeline,Linux CPU x64 NoContribops CI Pipeline,Linux GPU CI Pipeline,Linux GPU TensorRT CI Pipeline,MacOS CI Pipeline,MacOS NoContribops CI Pipeline,Windows CPU CI Pipeline,Windows GPU CI Pipeline,Windows GPU TensorRT CI Pipeline

SherlockNoMad commented:
/azp run orttraining-linux-ci-pipeline,orttraining-mac-ci-pipeline,orttraining-linux-gpu-ci-pipeline,centos7_cpu,Linux CPU Minimal Build E2E CI Pipeline,Linux Nuphar CI Pipeline,MacOS NoContribops CI Pipeline,Linux OpenVINO CI Pipeline,orttraining-distributed

SherlockNoMad commented:
sure. Appreciate it if you can contribute the CPU BatchNormGradient kernel !!

pranav-prakash commented:
@SherlockNoMad
Seems like maybe we can't remove `InsertMaxPoolOutput ` in `gradient_graph_builder.cc` after all? It fails the test
```
1: C++ exception with description "/onnxruntime_src/orttraining/orttraining/core/graph/gradient_builder_base.h:102 onnxruntime::training::ArgDef onnxruntime::training::GradientBuilderBase::O(size_t) const i < node_->OutputDefs().size() was false. 
```

although I'm not sure why this is since we register it in `graph_transformer_utils.cc` as well.

SherlockNoMad commented:
ah... ic.. the GradientCheckerTest.MaxPoolGrad is not invoking the GeneratePreTrainingTransformers, as it's taking the InferenceSession path to construct the test model. 

Please revert the Maxpool related changes, we will fix this in the latter PRs. Sorry for the inconvenience caused.

pranav-prakash commented:
@SherlockNoMad Done, thanks for determining the cause!

SherlockNoMad commented:
/azp run orttraining-linux-ci-pipeline,orttraining-mac-ci-pipeline,orttraining-linux-gpu-ci-pipeline,centos7_cpu,Linux CPU Minimal Build E2E CI Pipeline,Linux Nuphar CI Pipeline,MacOS NoContribops CI Pipeline,Linux OpenVINO CI Pipeline,orttraining-distributed

SherlockNoMad commented:
/azp run Windows WebAssembly CI Pipeline, orttraining-amd-gpu-ci-pipeline, orttraining-ortmodule, orttraining-ortmodule-distributed
